contracts-bedrock: StandardBridge pausability

tynes opened the pull request:
**Description**

Enable extended pause functionality in the standard bridge.
Includes updates to tests and the deploy script.

<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

maurelian committed:
contracts-bedrock: bridge pausability

Enable extended pause functionality in the standard bridge.
Includes updates to tests and the deploy script.

tynes committed:
contracts-bedrock: pause tests

tynes committed:
op-bindings: regenerate

tynes committed:
contracts-bedrock: bridge

tynes committed:
contracts-bedrock: regen lockfiles

tynes commented on packages/contracts-bedrock/src/L1/L1StandardBridge.sol:
```diff
@@ -69,12 +70,28 @@ contract L1StandardBridge is StandardBridge, ISemver {
+
+    /// @notice Initializes the contract.
+    /// @param _superchainConfig Address of the SuperchainConfig contract on this network.
+    function initialize(SuperchainConfig _superchainConfig) public initializer {
+        superchainConfig = _superchainConfig;
```
This is a good idea but we don't do it throughout the codebase. It is better to be standard

tynes commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -18,7 +18,7 @@ contract L1CrossDomainMessenger is CrossDomainMessenger, ISemver {
+    /// @notice Address of the SuperchainConfig contract.
     SuperchainConfig public superchainConfig;
 
     /// @notice Semantic version.
```
We know this is safe based on the `upgradeAndCall` pattern

maurelian commented on packages/contracts-bedrock/test/L1/L1StandardBridge.t.sol:
```diff
@@ -38,6 +39,31 @@ contract L1StandardBridge_Initialize_Test is Bridge_Initializer {
     }
 }
 
+contract L1StandardBridge_Pause_Test is Bridge_Initializer {
```
We need a test or tests to show that none of the withdrawal methods are callable when the bridge is paused.

tynes replied:
Good callout

maurelian commented on packages/contracts-bedrock/test/L1/L1StandardBridge.t.sol:
```diff
@@ -38,6 +39,31 @@ contract L1StandardBridge_Initialize_Test is Bridge_Initializer {
     }
 }
 
+contract L1StandardBridge_Pause_Test is Bridge_Initializer {
+    /// @dev Test that the accessors return the correct initialized values
+    function test_paused_succeeds() external {
+        assertEq(l1StandardBridge.paused(), superchainConfig.paused());
+    }
+
+    /// @dev Tests that the superchain config is called by the messengers paused function
+    function test_pause_callsSuperchainConfig_succeeds() external {
+        vm.expectCall(address(superchainConfig), abi.encodeWithSelector(SuperchainConfig.paused.selector));
+        l1StandardBridge.paused();
+    }
+
+    /// @dev Tests that changing the superchain config paused status changes the return value of the bridge.
+    function test_pause_matchesSuperchainConfig_succeeds() external {
+        assertFalse(l1StandardBridge.paused());
+        assertEq(l1StandardBridge.paused(), superchainConfig.paused());
+
+        vm.prank(superchainConfig.guardian());
+        superchainConfig.pause("identifier");
+
+        assertTrue(l1StandardBridge.paused());
+        assertEq(l1StandardBridge.paused(), superchainConfig.paused());
+    }
```
Thanks buddy, appreciate it.

tynes committed:
ctb: remove storage layout

tynes committed:
contracts-bedrock: test coverage for paused functionality

Only impacts deposits

tynes committed:
contracts-bedrock: update snapshots

tynes commented on packages/contracts-bedrock/test/L1/L1StandardBridge.t.sol:
```diff
@@ -38,6 +39,31 @@ contract L1StandardBridge_Initialize_Test is Bridge_Initializer {
     }
 }
 
+contract L1StandardBridge_Pause_Test is Bridge_Initializer {
+    /// @dev Test that the accessors return the correct initialized values
+    function test_paused_succeeds() external {
+        assertEq(l1StandardBridge.paused(), superchainConfig.paused());
+    }
+
+    /// @dev Tests that the superchain config is called by the messengers paused function
+    function test_pause_callsSuperchainConfig_succeeds() external {
+        vm.expectCall(address(superchainConfig), abi.encodeWithSelector(SuperchainConfig.paused.selector));
+        l1StandardBridge.paused();
+    }
+
+    /// @dev Tests that changing the superchain config paused status changes the return value of the bridge.
+    function test_pause_matchesSuperchainConfig_succeeds() external {
+        assertFalse(l1StandardBridge.paused());
+        assertEq(l1StandardBridge.paused(), superchainConfig.paused());
+
+        vm.prank(superchainConfig.guardian());
+        superchainConfig.pause("identifier");
+
+        assertTrue(l1StandardBridge.paused());
+        assertEq(l1StandardBridge.paused(), superchainConfig.paused());
+    }
```
Need to update this to pause withdrawals and not deposits

tynes replied:
This is close but not exactly how to perform withdrawals

tynes replied:
You were more correct than I expected. The only thing you did wrong was you used the legacy methods. We need to test both the legacy methods and the modern methods. The modern methods are implemented on the `StandardBridge` and are symmetric across L1 and L2. They are `finalizeBridgeETH` and `finalizeBridgeERC20`

tynes replied:
Your tests were very useful and helped me. They were an inspiration for my own.

tynes committed:
contracts-bedrock: standard bridge pausable withdrawals

tynes committed:
op-chain-ops: initialize L2StandardBridge in state

https://github.com/ethereum-optimism/optimism/pull/8377 is very
important because it would catch the regression of this not
being caught

tynes committed:
contracts-bedrock: locks

tynes committed:
op-bindings: regenerate

tynes committed:
lint: fix

tynes commented on packages/contracts-bedrock/semver-lock.json:
```diff
@@ -2,9 +2,9 @@
   "src/EAS/SchemaRegistry.sol": "0x5ee1a0c3b2bf1eb5edb53fb0967cf13856be546f0f16fe7acdc3e4f286db6831",
   "src/L1/DelayedVetoable.sol": "0x276c6276292095e6aa37a70008cf4e0d1cbcc020dbc9107459bbc72ab5ed744f",
-  "src/L1/L1CrossDomainMessenger.sol": "0xb154632221d578ac8af3f11eb74b296b959cbe6523ed6890a761cd614b1be79d",
+  "src/L1/L1CrossDomainMessenger.sol": "0x18b75f2ac0095aed2bb13d4352b5acfb19603b0c29040ed360ba70c491a09283",
```
Hmm why is the L1 xdm changing here?

maurelian commented on packages/contracts-bedrock/src/L2/L2StandardBridge.sol:
```diff
@@ -52,8 +52,8 @@ contract L2StandardBridge is StandardBridge, ISemver {
-    /// @custom:semver 1.5.0
-    string public constant version = "1.5.0";
+    /// @custom:semver 1.7.0
+    string public constant version = "1.7.0";
```
```suggestion
    string public constant version = "1.6.0";
```

tynes committed:
lockfiles: regenerate

tynes commented on packages/contracts-bedrock/src/L1/L1StandardBridge.sol:
```diff
@@ -69,12 +70,28 @@ contract L1StandardBridge is StandardBridge, ISemver {
-    /// @custom:semver 1.5.0
-    string public constant version = "1.5.0";
+    /// @custom:semver 2.0.0
+    string public constant version = "2.0.0";
```
Are you sure about this?

maurelian commented on packages/contracts-bedrock/test/L1/L1StandardBridge.t.sol:
```diff
@@ -38,6 +39,96 @@ contract L1StandardBridge_Initialize_Test is Bridge_Initializer {
+    }
+}
+
+contract L1StandardBridge_Pause_TestFail is Bridge_Initializer {
```
This test file pretty consistently adds `@dev` comments on the test functions, I think it's worth doing on these too.

maurelian replied:
@coderabbitai perhaps you could go through the diff in this file and suggest a natspec comment for each new test function? It should be in the format `/// @dev` and should provide a brief description of what is being tested.

maurelian replied:
Any chance you could format that as a patch file to make things easier for us?

maurelian replied:
no, the patch shouldn't delete any code.
Provide a patch that only adds the dev comments please.

tynes replied:
Thank you for the help. I have added the dev comments. In the future, it would be most useful to receive a commit suggestion as a comment so that I can add the patch directly through the ui

maurelian reviewed (approved):
Approving with one comment/suggestion that IMO should be addressed before merging.

tynes committed:
contracts-bedrock: update comments

tynes commented on packages/contracts-bedrock/test/L1/L1StandardBridge.t.sol:
```diff
@@ -38,6 +39,96 @@ contract L1StandardBridge_Initialize_Test is Bridge_Initializer {
+            _amount: 0,
+            _extraData: hex""
+        });
+    }
```
I did add the `@dev` comments in a future commit

tynes committed:
lint: fix
